sdk/web-pubsub/web-pubsub-client/src/webPubSubClient.ts

```diff
@@ -0,0 +1,1121 @@
+
+    // Auto reconnect or stop
+    if (this._options.autoReconnect) {
+      await this.autoReconnect();
+    } else {
+      await this.handleConnectionStopped();
+    }
+  }
+
+  private async autoReconnect(): Promise<void> {
+    let isSuccess = false;
+    let attempt = 0;
+    try {
+      while (!this._isStopping) {
+        try {
+          await this.startFromRestarting();
+          isSuccess = true;
+          break;
+        } catch (err) {
+          console.warn("An attempt to reconnect connection failed", err);
+
+          attempt++;
+          const delayInMs = this._reconnectRetryPolicy.nextRetryDelayInMs(attempt);
+
+          if (delayInMs == null) {
+            break;
+          }
+
+          try {
+            await delay(delayInMs);
+          } catch {}
+        }
+      }
+    } finally {
+      if (!isSuccess) {
+        this.handleConnectionStopped();
+      }
+    }
+  }
+
+  private handleConnectionStopped(): void {
+    this._isStopping = false;
+    this._state = WebPubSubClientState.Stopped;
+    this.emitEvent("stopped", {});
+  }
+
+  private async sendMessage(
+    message: WebPubSubMessage,
+    abortSignal?: AbortSignalLike
+  ): Promise<void> {
+    const payload = this._protocol.writeMessage(message);
+
+    if (!this._socket || this._socket.readyState !== WebSocket.OPEN) {
+      throw new Error("The connection is not connected.");
+    }
+    await sendAsync(this._socket, payload, abortSignal);
+  }
+
+  private async sendMessageWithAckId(
+    messageProvider: (ackId: number) => WebPubSubMessage,
+    ackId?: number,
+    abortSignal?: AbortSignalLike
+  ): Promise<WebPubSubResult> {
+    if (ackId == null) {
+      ackId = this.nextAckId();
+    }
+
+    const message = messageProvider(ackId);
+    if (!this._ackMap.has(ackId)) {
+      this._ackMap.set(ackId, new AckEntity(ackId));
+    }
+    const entity = this._ackMap.get(ackId)!;
+
+    try {
+      await this.sendMessage(message, abortSignal);
+    } catch (error) {
+      this._ackMap.delete(ackId);
+
+      let errorMessage: string = "";
+      if (error instanceof Error) {
+        errorMessage = error.message;
+      }
+      throw new SendMessageError(errorMessage, { ackId: ackId });
+    }
+
+    return await entity.promise();
+  }
+
+  private async handleConnectionClose(): Promise<void> {
+    // Clean ack cache
+    this._ackMap.forEach((value, key) => {
+      if (this._ackMap.delete(key)) {
+        value.reject(
+          new SendMessageError("Connection is disconnected before receive ack from the service", {
+            ackId: value.ackId,
+          } as SendMessageErrorOptions)
+        );
+      }
+    });
+
+    if (this._isStopping) {
+      console.warn("The client is stopping state. Stop recovery.");
+      this.handleConnectionCloseAndNoRecovery();
+      return;
+    }
+
+    if (this._lastCloseEvent && this._lastCloseEvent.code === 1008) {
+      console.warn("The websocket close with status code 1008. Stop recovery.");
```

Comment:
please look into `@azure/logger` instead of using `console.log` and friends directly